### PR TITLE
ccl/sqlproxyccl: directory proto and test server

### PR DESCRIPTION
**Pull request by darinpp**
* Defines a new interface between a tenant directory client and server
* Moves the tenant directory from the CC repo over
* Tenant directory modified to use the new interface
* Tenant directory modified to use stop.Stopper
* Modified pod watcher to use streaming grpc call
* Renamed package from directory to tenant, tenantID to ID etc
* Renamed references to k8s, pods to server, endpoints etc
* Prevents test tenant servers to start for non-existing/inactive tenants
* Adds ability to shut down individual tenant servers within a test cluster
* Adds a test directory server that can start/stop tenants
* Adds tests of the directory running agianst the test server
* Allow insecure connections from test tenant to KV server
* Fixed a race in kvtenantccl

Release note: None

### CODE REVIEW

**Review by andy-kimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64171)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @darinpp)

___
*[pkg/ccl/kvccl/kvtenantccl/connector.go, line 372 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0pMeW5_9Te3NlI4KX:-MZ0pMeW5_9Te3NlI4KY:bt0x51o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/kvccl/kvtenantccl/connector.go#L372)):*
> ```Go
> 			return nil, err
> 		}
> 		c.mu.RLock()
> ```

It'd be good to add a comment here explaining why the read lock is in fact needed.
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0rk8I0pw0vAIRSpCg:-MZ0rk8I0pw0vAIRSpCh:b-dkxqac) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L66)):*
> ```Go
> type Directory struct {
> 	// ctl is the directory client instance used to make directory server calls.
> 	ctl DirectoryClient
> ```

NIT: I think this should be renamed to `client` or `dir` rather than using `ctl` like in the operator.
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0qCz83ayoEKyfimTY:-MZ0qCz83ayoEKyfimTZ:b-8lmukt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L88)):*
> ```Go
> 
> // NewDirectory constructs a new Directory instance that tracks SQL tenant processes
> // managed by a given Directory server. The given context
> ```

NIT: spacing in comment.
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 257 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0rygDDjaHB8NQhNxa:-MZ0rygDDjaHB8NQhNxb:b48n2gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L257)):*
> ```Go
> // a notification and update the directory to reflect that change.
> func (d *Directory) watchEndpoints(ctx context.Context, stopper *stop.Stopper) error {
> ```

NIT: extra blank line - you're triggering my OCD : )
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 260 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0sh9p2LXqKY5sSSVf:-MZ0sh9p2LXqKY5sSSVg:bj1qbs2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L260)):*
> ```Go
> 	req := WatchEndpointsRequest{}
> 
> 	var waitInit sync.WaitGroup
> ```

Maybe add a comment here explaining why this WaitGroup is needed.
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0trMZ34YvT2d2YerM:-MZ0trMZ34YvT2d2YerN:b-c4et86) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L281)):*
> ```Go
> 					}
> 					log.Errorf(ctx, "err creating new watch endpoint client: %s", err)
> 					time.Sleep(time.Second)
> ```

Use the `sleepContext` helper method for this and any other sleeps, so that we cancel more quickly (and also consistently use it).
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 307 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0tDQ42KSIcFO2gEfk:-MZ0tDQ42KSIcFO2gEfl:bimw11f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L307)):*
> ```Go
> 			}
> 
> 			// Ensure that a directory entry exists for this tenant.
> ```

I think we decided to *not* create a new entry for pods that don't have a corresponding tenant in the cache yet. That would mean passing allowCreate=false below.
___
*[pkg/ccl/sqlproxyccl/tenant/directory.proto, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0uDJ4DrdSrt7vPvvD:-MZ0uDJ4DrdSrt7vPvvE:beh2vah) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.proto#L15)):*
> ```protobuf
> import "gogoproto/gogo.proto";
> 
> message WatchEndpointsRequest {}
> ```

This file needs comments on enums/messages/fields/etc.
___
*[pkg/ccl/sqlproxyccl/tenant/directory_test.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0uQSACxo4P2uqu1uN:-MZ0uQSACxo4P2uqu1uO:bvf6c57) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory_test.go#L30)):*
> ```Go
> )
> 
> func CreateTenant(tc serverutils.TestClusterInterface, id ID) error {
> ```

Is there some reason that these functions need to be exported? Also, I think it's nicer when helper code like this is at the bottom of the file, so that the actual tests can be placed at the top of the file.
___
*[pkg/ccl/sqlproxyccl/tenant/entry.go, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0v17vDB3wfnBQeCIo:-MZ0v17vDB3wfnBQeCIp:bx6wve9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/entry.go#L33)):*
> ```Go
> 	TenantID ID
> 
> 	// Full name of the tenant's cluster i.e. dim-dog
> ```

NIT: missing period at end of comment.
___
*[pkg/ccl/sqlproxyccl/tenant/entry.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0vNI-81kPE2PDOhGn:-MZ0vNI-81kPE2PDOhGo:b-42okq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/entry.go#L68)):*
> ```Go
> // that in the entry. After this is called once, all future calls return the
> // same result (and do nothing).
> func (e *tenantEntry) Initialize(ctx context.Context, ctl DirectoryClient) error {
> ```

NIT: similar to above, rename `ctl` to `client` or `dir`.
___
*[pkg/ccl/sqlproxyccl/tenant/entry.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0wPG92pE2eoogm08r:-MZ0wPG92pE2eoogm08s:b-1ngv5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/entry.go#L192)):*
> ```Go
> // endpoints available rather than blocking.
> func (e *tenantEntry) ensureTenantEndpoint(
> 	ctx context.Context, ctl DirectoryClient, errorIfNoEndpoints bool,
> ```

NIT: rename ctl here as well (and any other places I missed).
___
*[pkg/ccl/sqlproxyccl/tenant/entry.go, line 232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0wVOp6mfs3h3fSrpy:-MZ0wVOp6mfs3h3fSrpz:bo2bhcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/entry.go#L232)):*
> ```Go
> 		// Get endpoint information for the newly resumed tenant. Except in rare race
> 		// conditions, this is expected to immediately find an IP address, since
> 		// the above call started a tenant process that  already has an IP address.
> ```

NIT: double space in comment.
___
*[pkg/ccl/sqlproxyccl/tenant/entry.go, line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0wbql9__QhDPT3WrP:-MZ0wbql9__QhDPT3WrQ:bc9ytti) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/entry.go#L259)):*
> ```Go
> 	ctx context.Context, ctl DirectoryClient,
> ) (ips []string, err error) {
> 	// List endpoints with a "tenant-id" label that matches this tenant's ID.
> ```

NIT: This comment is out-of-date.
___
*[pkg/ccl/sqlproxyccl/tenant/tenant_id.go, line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0wvBP08dKDeUock-2:-MZ0wvBP08dKDeUock-3:byh869o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/tenant_id.go#L26)):*
> ```Go
> // For more information, see:
> //   https://github.com/cockroachdb/cockroach/blob/release-20.2/pkg/roachpb/tenant.go#L19-L24
> type ID uint64
> ```

We should use `roachpb.TenantID` instead of defining a new variant in CRDB.
___
*[pkg/ccl/sqlproxyccl/tenant/test_directory_svr.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0xH8fDABhr6xdOOOl:-MZ0xH8fDABhr6xdOOOm:bx6wve9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/test_directory_svr.go#L73)):*
> ```Go
> 
> 	// When both mutexes need to be held, the locking should always be
> 	// proc first and listen second
> ```

NIT: missing period at end of comment.
___
*[pkg/ccl/sqlproxyccl/tenant/test_directory_svr.go, line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0xrul6rmH06LVC7AW:-MZ0xrul6rmH06LVC7AX:b-5pku5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/test_directory_svr.go#L273)):*
> ```Go
> 		return nil, err
> 	}
> 	process := &Process{SQL: sql.Addr()}
> ```

What happens if the test is stopped using ctrl-C? Will the processes still be cleaned up?
___
*[pkg/testutils/serverutils/test_server_shim.go, line 225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0yCGDDoRD4sQq8OsX:-MZ0yCGDDoRD4sQq8OsY:b-9zktle) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/testutils/serverutils/test_server_shim.go#L225)):*
> ```Go
> 	// StartTenant spawns off tenant process connecting to this TestServer with
> 	// context
> 	StartTenantWithCtx(ctx context.Context, params base.TestTenantArgs) (TestTenantInterface, error)
> ```

Why not just update `StartTenant` and then change all callers to pass a context? I don't see much value in having both variants.

**Review by darinpp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64171)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball)

___
*[pkg/ccl/kvccl/kvtenantccl/connector.go, line 372 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0pMeW5_9Te3NlI4KX:-MZK3Xzh5QkkOHvmiifa:by6z1x2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/kvccl/kvtenantccl/connector.go#L372)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

It'd be good to add a comment here explaining why the read lock is in fact needed.
</blockquote></details>

I refactored it a bit as there was another issue with this code. Now all paths lock when accessing the client. We generally don't put comments justifying locking if there is access from separate threads to the same field. This is the case here so I guess a comment would be appropriate only of the locks are not being obtained (as it was before).
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0rk8I0pw0vAIRSpCg:-MZK4HkHAzQH1rSleGrK:b-p3vt1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L66)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: I think this should be renamed to `client` or `dir` rather than using `ctl` like in the operator.
</blockquote></details>

renamed to `client`
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0qCz83ayoEKyfimTY:-MZK4JwA89eauw631unP:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L88)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: spacing in comment.
</blockquote></details>

fixed
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 257 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0rygDDjaHB8NQhNxa:-MZK4LZPDBLliW-FBzbC:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L257)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: extra blank line - you're triggering my OCD : )
</blockquote></details>

fixed
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 260 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0sh9p2LXqKY5sSSVf:-MZK4MU47t76FCfdLFpy:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L260)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Maybe add a comment here explaining why this WaitGroup is needed.
</blockquote></details>

added
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0trMZ34YvT2d2YerM:-MZK4Nfp8Bx6vzqOsAAe:b-n9tkhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L281)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Use the `sleepContext` helper method for this and any other sleeps, so that we cancel more quickly (and also consistently use it).
</blockquote></details>

changed
___
*[pkg/ccl/sqlproxyccl/tenant/directory.go, line 307 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0tDQ42KSIcFO2gEfk:-MZK4R3g5zOlGyDPBEo6:b7l9y9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.go#L307)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think we decided to *not* create a new entry for pods that don't have a corresponding tenant in the cache yet. That would mean passing allowCreate=false below.
</blockquote></details>

OK. I fixed to not create entry. This also required a change in one of the tests.
___
*[pkg/ccl/sqlproxyccl/tenant/directory.proto, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0uDJ4DrdSrt7vPvvD:-MZK4Xar7JwloDEY3li4:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory.proto#L15)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This file needs comments on enums/messages/fields/etc.
</blockquote></details>

Added.
___
*[pkg/ccl/sqlproxyccl/tenant/directory_test.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0uQSACxo4P2uqu1uN:-MZK4_vGDZ5uPYwnCUv-:b-hvx29h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/directory_test.go#L30)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Is there some reason that these functions need to be exported? Also, I think it's nicer when helper code like this is at the bottom of the file, so that the actual tests can be placed at the top of the file.
</blockquote></details>

No good reason. I changed them so they are not exported anymore.
___
*[pkg/ccl/sqlproxyccl/tenant/entry.go, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0v17vDB3wfnBQeCIo:-MZK4jp_-3tR79Qin8ER:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/entry.go#L33)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: missing period at end of comment.
</blockquote></details>

fixed
___
*[pkg/ccl/sqlproxyccl/tenant/entry.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0vNI-81kPE2PDOhGn:-MZK4kzi1TRFaVCoFt0I:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/entry.go#L68)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: similar to above, rename `ctl` to `client` or `dir`.
</blockquote></details>

done
___
*[pkg/ccl/sqlproxyccl/tenant/entry.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0wPG92pE2eoogm08r:-MZK4mNr-KEBBO7-Lhhd:b-6qdckw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/entry.go#L192)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: rename ctl here as well (and any other places I missed).
</blockquote></details>

all changed
___
*[pkg/ccl/sqlproxyccl/tenant/entry.go, line 232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0wVOp6mfs3h3fSrpy:-MZK4nrr-0T2IzNfOPA5:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/entry.go#L232)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: double space in comment.
</blockquote></details>

fixed
___
*[pkg/ccl/sqlproxyccl/tenant/entry.go, line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0wbql9__QhDPT3WrP:-MZK4ovP08zayE-G6HYS:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/entry.go#L259)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: This comment is out-of-date.
</blockquote></details>

fixed
___
*[pkg/ccl/sqlproxyccl/tenant/tenant_id.go, line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0wvBP08dKDeUock-2:-MZK4qVm0f6wTn-QyLXV:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/tenant_id.go#L26)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

We should use `roachpb.TenantID` instead of defining a new variant in CRDB.
</blockquote></details>

done
___
*[pkg/ccl/sqlproxyccl/tenant/test_directory_svr.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0xH8fDABhr6xdOOOl:-MZK4srz1lFDis8P5uR5:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/test_directory_svr.go#L73)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: missing period at end of comment.
</blockquote></details>

fixed
___
*[pkg/ccl/sqlproxyccl/tenant/test_directory_svr.go, line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0xrul6rmH06LVC7AW:-MZK6ltZCUkQtEzb2YuL:b1pjhua) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/ccl/sqlproxyccl/tenant/test_directory_svr.go#L273)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

What happens if the test is stopped using ctrl-C? Will the processes still be cleaned up?
</blockquote></details>

The child processes won't be killed when the main process is killed. I'll add something in another PR that will be OS specific most likely to take care of this.
___
*[pkg/testutils/serverutils/test_server_shim.go, line 225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0yCGDDoRD4sQq8OsX:-MZK76cL8VNDtbJ5ar9W:beqlhdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c01679c0d20f5d15d9d2077fa1bb1cc16b1a88/pkg/testutils/serverutils/test_server_shim.go#L225)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Why not just update `StartTenant` and then change all callers to pass a context? I don't see much value in having both variants.
</blockquote></details>

Agree. Fixed.

File: pkg/ccl/kvccl/kvtenantccl/connector.go [hunk 1]
```diff
@@ -365,12 +365,17 @@ func (c *Connector) getClient(ctx context.Context) (roachpb.InternalClient, erro
 		dialCtx := c.AnnotateCtx(context.Background())
 		dialCtx, cancel := c.rpcContext.Stopper.WithCancelOnQuiesce(dialCtx)
 		defer cancel()
-		err := c.rpcContext.Stopper.RunTaskWithErr(dialCtx, "kvtenant.Connector: dial", c.dialAddrs)
+		var client roachpb.InternalClient
```
Comment (darinpp) [inline pkg/ccl/kvccl/kvtenantccl/connector.go:368]: @nvanbenschoten take a look and let me know if this will take care of the locking problem.

**Review by andy-kimball — APPROVED** [review]
:lgtm:

All the directory stuff looks good to me. It'd be good to get another reviewer for some of the scaffolding code like the connector, stopping, testing, etc., which I'm not too familiar with.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-:-MZKB5BL6NnBApiw2jLx:b-4ejg16)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @darinpp, and @nvanbenschoten)

___
*[pkg/ccl/sqlproxyccl/tenant/directory.proto, line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZKAbH_7_YRwDQR5t85:-MZKAbH_7_YRwDQR5t86:bi9nzom) ([raw file](https://github.com/cockroachdb/cockroach/blob/c90caeb0f508b24df457f519c6c93d8ed9ef7299/pkg/ccl/sqlproxyccl/tenant/directory.proto#L40)):*
> ```protobuf
> // ListEndpointsRequest is used to query the server for the list of current endpoints of a given tenant.
> message ListEndpointsRequest {
>   // tenant id identifies the tenant for which the client is requesting a list of the endpoints.
> ```

NIT: All of the field comments should be capitalized in Go-style (e.g. `TenantID`) because these will be inserted into the generated Go code after being converted from snake case.

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64171)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @darinpp)

___
*[pkg/ccl/kvccl/kvtenantccl/connector.go, line 368 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0Hqj2-ByNDNo253Di-r2-368:-MZLehNxBmy6vcpGJoLX:but3nxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c90caeb0f508b24df457f519c6c93d8ed9ef7299/pkg/ccl/kvccl/kvtenantccl/connector.go#L368)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

@nvanbenschoten take a look and let me know if this will take care of the locking problem.
</blockquote></details>

This is close, but I think we still want to set `c.mu.client` in this singleflight function, instead of having each goroutine do so after the result has been broadcasted back out. We just don't want to lock, set it, unlock, and then read it again before returning it, lest it be reset by someone else during the period where we weren't holding the lock.

So I think we'll want to replace the old:
```go
// NB: read lock not needed.
return c.mu.client, nil
```
With:
```go
c.mu.Lock()
defer c.mu.Unlock()
c.mu.client = client
return client, nil
```

And then remove the change below in the `case res := <-ch:` block.

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64171)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @darinpp)

___
*[pkg/ccl/kvccl/kvtenantccl/connector.go, line 368 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64171#-MZ0Hqj2-ByNDNo253Di-r2-368:-MZT1yHN4At9fkgKPyGk:bi5ui57) ([raw file](https://github.com/cockroachdb/cockroach/blob/c90caeb0f508b24df457f519c6c93d8ed9ef7299/pkg/ccl/kvccl/kvtenantccl/connector.go#L368)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is close, but I think we still want to set `c.mu.client` in this singleflight function, instead of having each goroutine do so after the result has been broadcasted back out. We just don't want to lock, set it, unlock, and then read it again before returning it, lest it be reset by someone else during the period where we weren't holding the lock.

So I think we'll want to replace the old:
```go
// NB: read lock not needed.
return c.mu.client, nil
```
With:
```go
c.mu.Lock()
defer c.mu.Unlock()
c.mu.client = client
return client, nil
```

And then remove the change below in the `case res := <-ch:` block.
</blockquote></details>

This part :lgtm: now. Thanks for cleaning it up.

**Pull request merged**